README
                             Locale::KeyedText
                                  README
---------------------------------------------------------------------------

by Darren Duncan <darren@DarrenDuncan.net>

PREFACE

This distribution features the Perl module named "Locale::KeyedText"; see
the file lib/Locale/KeyedText.pm for the main documentation, which is in
POD format, plus the Dependencies and Copyright.  The distribution also
includes a Changes file to say what changes already happened, and a TODO
file to say what changes have yet to happen.  The file INSTALL should help
with installation issues.  The examples/ directory contains several small
programs you can try out and see what Locale::KeyedText can do.  Read below
for how to get support or keep up to date.

KEEPING UP TO DATE

My module set is constantly under development.  The canonical copies are
all linked to through my personal website, "http://www.darrenduncan.net/",
or through my business website, "https://muldis.com".

For those of you that are interested in the bleeding edge of development, I
have all parts of the Locale::KeyedText project in one or more public
repositories.  The Perl Locale::KeyedText project is in a Git repository
whose public clone url is "git://github.com/muldis/Locale-KeyedText.git" and
whose web information url is at "https://github.com/muldis/Locale-KeyedText".
The Raku Locale::KeyedText project used to be in a Subversion repository for
the Raku Pugs project but the host domain pugscode.org has since expired.

You can download slightly older versions of all of my modules from the
public CPAN archives at "http://www.cpan.org/authors/id/D/DU/DUNCAND/", or
on my own web server at "http://www.darrenduncan.net/d/perl/".

CPAN also extracts the POD from modules and maintains that documentation in
an organized and searchable fashion.  Their main search site is at
"http://search.cpan.org", and my own DUNCAND author page is at
"http://search.cpan.org/search?author=DUNCAND".

The main CPAN search site url for Locale::KeyedText for Perl is
"http://search.cpan.org/dist/Locale-KeyedText/"; that is the url you should
link to presently if you want something to consider the "home page" of
Locale::KeyedText; it is best to refer people there by default.

You can download my oldest tarred and gzipped distributions from BACKPAN,
at "http://history.perl.org/backpan/authors/id/D/DU/DUNCAND/", or from my
web server, at "http://www.darrenduncan.net/d/perl/archives/", right back
to the libdwg 1.0, first released on 2000 July 23.

SUPPORT

Currently I don't have any support arranged with other people, lists,
newsgroups, or otherwise.  Feel free to ask me if you can't figure things
out on your own, or another person whom you know has used this.  I may
start a mailing list for support issues later, so that users of this module
can help each other with it.

FIN

Share and Enjoy!


Changes
                             Locale::KeyedText
                                  Changes
---------------------------------------------------------------------------

20xx-xx-xx

    * Locale::KeyedText version 2.1.2 for Perl is released on CPAN as
    Locale-KeyedText-2.1.2.tar.gz.

    * Make the distro pass the Perl Kwalitee "has_abstract_in_pod" test:
    For every pod file, tweak its NAME section so the abstract
    is on the same line as the package name rather than the next line.

    * TODO: meta_yml_has_provides

2019-10-20

    * Locale::KeyedText version 2.1.1 for Perl is released on CPAN as
    Locale-KeyedText-2.1.1.tar.gz.

    * Between the prior release and this one, the programming language
    formerly known as "Perl 6" was renamed to "Raku".  As such, all
    documentation that refers to that language was updated to use its new
    name, and all language that referred to "Perl 5" now simply refers to
    "Perl" because "Perl" now unambiguously means Perl 5.  The only
    exception to these changes is in documentation about history, such as
    all prior entries in this Changes file.

    * Lowered the declared minimum Perl version dependency to 5.8.0 (2002)
    from 5.8.1 (2003) since the prior .1 requirement seemed rather
    arbitrary and all 5.8.z and newer should suffice.

    * The declared minimum version dependencies for some Perl-bundled
    modules were lowered to those shipped with 5.8.0 from those with 5.8.1:
    ExtUtils::MakeMaker is now 6.03 from 6.17,
    "overload" is now 1.00 from 1.01.

    * As long as your Perl is version 5.8.1 (2003) or newer, each of the
    Locale::KeyedText classes requires only the Perl core and core modules
    to run, and nothing else, so usage is trivially easy. If your Perl is
    instead version 5.8.0 (2002), you need only also install the 1
    dual-life CPAN module Scalar::Util version 1.13+; 1.09 is the actual
    minimum dependency, but 1.13 matches what was bundled with Perl 5.8.1.

    * Locale::KeyedText is now released under the terms of the Artistic
    License version 2 rather than the GNU Lesser General Public License
    (LGPL) version 3+.

    * Updated all documentation to use an updated canonical Muldis Data
    Systems url of https://muldis.com rather than several http:// variants.

    * Incremented all copyright year range-ends to 2019.

2016-08-06

    * Locale::KeyedText version 2.1.0 for Perl 5 is released on CPAN as
    Locale-KeyedText-2.1.0.tar.gz.

    * The sole change of this release is to rearrange the lib/* files at a
    high level, mainly by splitting those files apart into smaller files.

    * Added new file lib/Locale/KeyedText.pod which contains all the POD
    documentation that was in KeyedText.pm and en.pm.  As such, all of the
    lib/*.pm files now only contain Perl code, and all POD now lives in
    lib/*.pod files.  The first primary benefit of this change is that the
    code and documentation can more easily be organized differently where
    we want, rather than their needing a 1:1 structural correspondance.
    The second primary benefit of this change is that now all POD will be
    automatically HTMLized by GitHub or other repositories for easier
    reading, rather than POD inside *.pm files not getting that treatment.

    * Split the file lib/Locale/KeyedText.pm into itself plus 2 new files
    lib/Locale/KeyedText/Message.pm and lib/Locale/KeyedText/Translator.pm.
    The end result is that this distribution no longer has any primary
    packages that are "cuckoo packages", meaning packages declared in a
    file with a different name than the package.  While such bundling can
    be a reasonable design choice, the CPAN infrastructure does run into
    some problems where those exist.  For example, one can ask a CPAN
    client to "install" such a package which it will then do but then treat
    as if it hadn't.  Therefore, in the future, "cuckoo packages" should be
    restricted more to ones that are truly internal and not part of a
    public API of the module; Message and Translator ARE part of the API.

    * This release has no other changes, and in particular no functional
    changes.  However, its version is being incremented by 0.1.0 rather
    than 0.0.1 because it breaks forwards compatibility; one can now "use"
    Message or Translator directly, and previously they couldn't.

2016-08-05

    * Locale::KeyedText version 2.0.0 for Perl 5 is released on CPAN as
    Locale-KeyedText-2.0.0.tar.gz.

    * This is the first release of Locale::KeyedText after a full decade
    with no development, and it has a new major version largely due to the
    module having gotten a number of large incompatible changes since the
    early 1.x releases (particularly in release 1.06_3).

    * New code file versions are: KeyedText.pm 2.0.0 and en.pm 2.0.0.

    * The first main purpose of this release is to excise all of
    Locale::KeyedText's runtime dependencies on non-Perl-core modules, so
    that it can be installed and used with a relative minimum of overhead.
    As of this release, each of the 2 Locale::KeyedText classes requires
    only the Perl core or core modules to run, and nothing else.  For the
    test suite that is also the case, at least if that Perl is version
    5.10.1 or newer; otherwise, the test suite also depends on just the 1
    dual-life CPAN module Test::More version 0.92 or newer.  This release
    is also about bringing other aspects of modernity to this distribution.

    * Note that despite the significant internal refactoring or
    reimplementation of Locale::KeyedText, release 2.0.0 should be entirely
    backwards-compatible; any code written for release 1.72.0+, or even
    1.70.0+, should work with it unchanged.

    * This distribution has excised its direct external dependencies on the
    modules Class::Std, Class::Std::Utils, Readonly, "only" and "version".
    The dependencies on those were mostly added with release 1.06_3 per

Makefile.PL
use 5.008000;
use utf8;
use strict;
use warnings;

use ExtUtils::MakeMaker;

# Make setting optional MakeMaker parameters more readable.
sub OPTIONAL {
   return () unless $ExtUtils::MakeMaker::VERSION ge shift;
   return @_;
}

WriteMakefile(
   NAME => 'Locale::KeyedText',
   VERSION => '2.1.1',
   ABSTRACT => 'Refer to user messages in programs by keys',
   AUTHOR => 'Darren Duncan <darren@DarrenDuncan.net>',
   PREREQ_PM => {
      # This list is a copy of "META_ADD:[configure_|build_|]requires"
      # where the other is canonical; this list is for use by MakeMaker
      # itself, which in general wouldn't see the info in the other lists,
      # as the other lists are primarily used when generating the META.yml/json.
      'ExtUtils::MakeMaker' => '6.03',
      'overload' => '1.00',
      'Scalar::Util' => '1.13',
      'Test::More' => '0.92',
   },
   OPTIONAL( '6.48',
      MIN_PERL_VERSION => '5.008000',
   ),
   OPTIONAL( '6.31',
      LICENSE => 'artistic_2',
   ),
   OPTIONAL( '6.46',
      # Use META_ADD instead of META_MERGE so that we can remove
      # any build-time dependencies that MakeMaker will put into
      # the requires field.  (Comment From DBD::SQLite; do *we* need that?)
      META_ADD => {
         configure_requires => {
            'ExtUtils::MakeMaker' => '6.03', # bundled with Perl 5.8.0
         },
         build_requires => {
            'ExtUtils::MakeMaker' => '6.03', # bundled with Perl 5.8.0
            # These are bundled with newer versions of Perl only.
            'Test::More' => '0.92', # bundled with Perl 5.10.1, min 0.88 needed
         },
         requires => {
            'perl' => '5.008000', # as MIN_PERL_VERSION no affect META.yml
            'overload' => '1.00', # bundled with Perl 5.8.0
            # These are bundled with newer versions of Perl only.
            'Scalar::Util' => '1.13', # bundled with Perl 5.8.1, min 1.09 needed
         },
      },
      META_MERGE => {
         no_index => {
            directory => ['examples','t'],
         },
         resources => {
            homepage => 'https://muldis.com',
            license => 'https://www.perlfoundation.org/artistic-license-20.html',
            MailingList => 'http://mm.darrenduncan.net/mailman/listinfo',
            repository => 'https://github.com/muldis/Locale-KeyedText',
         },
      },
   ),
);

1;
